kvs/client/python/anna/ipc_client.py

```diff
@@ -12,59 +12,167 @@
 #  See the License for the specific language governing permissions and
 #  limitations under the License.
 
-GET_ADDR = "ipc:///requests/get"
-PUT_ADDR = "ipc:///requests/put"
+GET_REQUEST_ADDR = "ipc:///requests/get"
+PUT_REQUEST_ADDR = "ipc:///requests/put"
 
+GET_RESPONSE_ADDR_TEMPLATE = "ipc:///responses_%d/get"
+PUT_RESPONSE_ADDR_TEMPLATE = "ipc:///responses_%d/put"
+
+TIMEOUT = 5000
+
+from .functions_pb2 import *
 from .kvs_pb2 import *
 from .lattices import *
 import zmq
 
 class IpcAnnaClient:
-    def __init__(self, offset=0):
+    def __init__(self, thread_id = 0, offset=0):
         self.context = zmq.Context(1)
 
-        self.get_socket = self.context.socket(zmq.REQ)
-        self.get_socket.connect(GET_ADDR)
+        self.get_push_socket = self.context.socket(zmq.PUSH)
+        self.get_push_socket.connect(GET_REQUEST_ADDR)
+
+        self.put_push_socket = self.context.socket(zmq.PUSH)
+        self.put_push_socket.connect(PUT_REQUEST_ADDR)
+
+        self.get_pull_socket = self.context.socket(zmq.PULL)
+        self.get_pull_socket.bind(GET_RESPONSE_ADDR_TEMPLATE % thread_id)
+        self.get_pull_socket.RCVTIMEO = TIMEOUT
 
-        self.put_socket = self.context.socket(zmq.REQ)
-        self.put_socket.connect(PUT_ADDR)
+        self.put_pull_socket = self.context.socket(zmq.PULL)
+        self.put_pull_socket.bind(PUT_RESPONSE_ADDR_TEMPLATE % thread_id)
+        self.put_pull_socket.RCVTIMEO = TIMEOUT
 
-    def get(self, key, ltype):
+    def get(self, refs):
         request = KeyRequest()
         request.type = GET
 
-        tp = request.tuples.add()
-        tp.lattice_type = ltype
-        tp.key = key
+        deserialize = {}
+
+        for ref in refs:
+            tp = request.tuples.add()
+            tp.key = ref.key
+            tp.lattice_type = ref.obj_type
+
+            deserialize[ref.key] = ref.deserialize
 
-        self.get_socket.send(request.SerializeToString())
+            if ref.obj_type == CROSSCAUSAL:
+                print("Error: found cross causal lattice type in non causal mode!")
 
-        resp = KeyResponse()
-        resp.ParseFromString(self.get_socket.recv())
+        request.response_address = GET_RESPONSE_ADDR_TEMPLATE % thread_id
 
-        tp = resp.tuples[0]
+        self.get_push_socket.send(request.SerializeToString())
 
-        if tp.error == 1:
-            print('Key %s does not exist!' % (key))
+        try:
+            msg = self.get_pull_socket.recv()
+        except zmq.ZMQError as e:
+            if e.errno == zmq.EAGAIN:
+                print("request timed out")
```

Comment:
This should be logged instead of being printed.

Reply:
fixed